packages/http/src/validator/validators/body.ts

```diff
@@ -1,82 +1,138 @@
 import { IPrismDiagnostic } from '@stoplight/prism-core';
 import { IMediaTypeContent } from '@stoplight/types';
-import { get } from 'lodash';
-import { parse as deepParseUrlEncoded } from 'qs';
+import { get, partial } from 'lodash';
 import * as typeIs from 'type-is';
 import { body } from '../deserializers';
 
 import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
+import { IHttpEncoding } from '@stoplight/types/dist';
+import * as Array from 'fp-ts/lib/Array';
+import * as Either from 'fp-ts/lib/Either';
+import * as Option from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
 import { JSONSchema } from '../../types';
-import { validateAgainstSchema } from '../validators/utils';
 import { IHttpValidator } from './types';
+import { validateAgainstSchema } from './utils';
 
 export class HttpBodyValidator implements IHttpValidator<any, IMediaTypeContent> {
-  constructor(private _prefix: string) {}
+  constructor(private prefix: string) {}
 
   public validate(target: any, specs: IMediaTypeContent[], mediaType?: string): IPrismDiagnostic[] {
-    const { _prefix: prefix } = this;
-    const content = this.getContent(specs, mediaType);
-
+    const content = getContent(specs, mediaType);
     const schema = get(content, 'schema');
 
-    if (!schema) {
+    try {
+      return pipe(
+        schema ? Either.right(target) : Either.left([]),
+        Either.chain(partial(maybeValidateFormBody, schema!, content, mediaType)),
+        Either.chain(partial(validateBody, schema!)),
+        Either.fold(partial(applyPrefix, this.prefix), () => []),
+      );
+    } catch (e) {
+      require('fs').appendFileSync('myk.txt', JSON.stringify(e.stack));
       return [];
     }
+  }
+}
 
-    if (mediaType && typeIs.is(mediaType, 'application/x-www-form-urlencoded')) {
-      const shallowParsedTarget = shallowParseUrlEncoded(target);
-      target = deepParseUrlEncoded(target);
-
-      const encodings = get(content, 'encodings', []);
-      for (const encoding of encodings) {
-        const allowReserved = get(encoding, 'allowReserved', false);
-        const property = encoding.property;
-        const value = shallowParsedTarget[property];
-
-        if (!allowReserved && typeof value === 'string' && value.match(/[\/?#\[\]@!$&'()*+,;=]/)) {
-          return [
-            {
-              path: [prefix, property],
-              message: 'Reserved characters used in request body',
-              severity: DiagnosticSeverity.Error,
-            },
-          ];
-        }
+function validateBody(schema: JSONSchema, target: any) {
```

Comment:
can `target` be something more concrete?

Reply:
At this point, it can be `string|object`. I guess. I'm quite sure that at some point we should decide how `body` is transported around prism infrastructure. Unfortunately, it means some refactor which is out of this issue's scope.